File: Data Structures & Algorithms/invert-a-binary-tree/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
   public:
    TreeNode* invertTree(TreeNode* root) {
        // Return nullptr if the root is nullptr
        if (root == nullptr) {
            return nullptr;
        }

        // Swap the left and right children node
        swap(root->left, root->right);

        // Recursive call for each left and right node
        invertTree(root->left);
        invertTree(root->right);

        // Return the modified tree's root node
        return root;
    }
};

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        # Return empty if root is empty
        if not root:
            return []

        result = []
        q = deque([root])  # Push root so the while loop can start

        # Exist when the queue is empty, which means that we've processed everything
        while q:
            # Cache the current level size from the queue
            # Note that we have to do this because q's size changes in the for loop below
            level_size = len(q)
            level = []

            # Iterate through the queue to process current level
            for _ in range(level_size):
                # Pop the front node and push its value into level
                front = q.popleft()
                level.append(front.val)

                # Push front node's children into queue if they exist
                if front.left:
                    q.append(front.left)
                if front.right:
                    q.append(front.right)

            # Push level into result because the level container
            # should now contain current level's nodes in right order
            result.append(level)
        return result
